### ISSUE TITLE
Option to disable looping

### ISSUE THREAD

**Issue by sodapopcan** [AndrewRadev/sideways.vim#56]
Hey Andrew,

Would you be amenable to including an option to disable looping?  I would like to build functionality that is actually a combination of two of your plugins though it would require the result of `sideways#MoveLeft()` and `sideways#MoveRight()` to return zero if they are on the first or last arguments.  I'm happy to take on the work myself if so.

On a different note, thank you so much for all your great plugins.  I am a long time user of splitjoin (looks like I added it back in 2015) but for some reason didn't branch out into your other work much until recently.  Deleft, whitespaste, and sideways are all amazing!  Perhaps you have other gold but have not looked yet.  I also enjoy your YouTube videos.  In any event, thanks for all your hard work!

Andrew

**Comment by AndrewRadev** [comment]
Thank you very much for your kind words :). I'd like to think I've got other good ones, yeah, but I'm particularly glad to see somebody putting together their own setup by using their components.

I've pushed a commit that adds the ability to disable looping around. I've done this on a setting level, since I think it could be generally preferable for some users, but for your use case, you should be able to run `splitjoin#MoveLeft({ 'loop': 0 })` to get the non-looping behaviour on a case-by-case basis.

Could you try it and let me know if it works as you'd expect?

**Comment by sodapopcan** [comment]
Whoa, thank you so much, that was very fast :)

Unfortunately I'm getting an "undefined variable `opts`" error.  It looks like you have an opts [here](https://github.com/AndrewRadev/sideways.vim/commit/bde47cd3b058b0cfffbedc62529f713bf0c4de33#diff-f3c99b2ef404b227897a7594d1f65713ced0a78579480c2600cdf4cb39552157R21) (and it's corresponding MoveRight version) that should be `a:1`.  I have changed these but am now getting:

```
Error detected while processing function <SNR>125_arg_left[1]..sideways#MoveLeft[3]..sideways#Parse[16]..sideways#parsing#Parse[13]..<SNR>178_ParseItems:                                                                                                     
line  108:                                                                                                                                                                                                                                                    
E523: Not allowed here
```

I'm looking into it further but otherwise from looking at the code this does exactly what I need it to.  Thank you so much!

**Comment by sodapopcan** [comment]
Sorry, I was testing that improperly.  Changing the `opts` to `a:1` fixes it.

**Comment by AndrewRadev** [comment]
Whoops, tested that everything still works, but forgot to test that providing the dict as arguments works 😅. Pushed a fix.

**Comment by sodapopcan** [comment]
Thank you!

If you're interested I write Elixir and was making a thing to shift args as args but also as pipes.  More of a neat party trick but I was actually using it today, though I think I'm going to need to do a bit of work for splitjoin for Elixir as it doesn't always work.  I also may eventually give in an use NeoVim with treesitter, lol, I dunno.

Here's the code if you're interested.  Could use a bit of a refactor probably, I only did one refactor pass so far.

```viml
function! s:define_argument_mappings()
  if exists('g:loaded_sideways')
    omap aa <Plug>SidewaysArgumentTextobjA
    xmap aa <Plug>SidewaysArgumentTextobjA
    omap ia <Plug>SidewaysArgumentTextobjI
    xmap ia <Plug>SidewaysArgumentTextobjI

    nnoremap <aa :call <sid>arg_left(0)<cr>
    nnoremap >aa :call <sid>arg_right(0)<cr>
    nnoremap <ia :call <sid>arg_left(1)<cr>
    nnoremap >ia :call <sid>arg_right(1)<cr>
  endif
endfunction

function! s:arg_left(inner)
  if a:inner
    call sideways#MoveLeft()
  elseif !sideways#MoveLeft({'loop': 0})
    call sj#Split()
  endif
endfunction

function! s:arg_right(inner)
  if a:inner
    call sideways#MoveRight()
  elseif sideways#MoveRight({'loop': 0})
    return
  endif

  let curr_line = getline('.')
  let prev_line = getline(line('.') - 1)
  let next_line = getline(line('.') + 1)
  let pipe_start = '^\%(\s*\)\?|>'

  let on_func = curr_line =~ pipe_start && prev_line !~ pipe_start
  let on_pipe_arg = curr_line !~ pipe_start && next_line =~ pipe_start

  if on_pipe_arg || on_func 
    if on_pipe_arg
      normal! j
    endif

    normal! 0t(
    let arg_pos = [line('.') - 1, col('.') - 1]

    call sj#Join()
    call cursor(arg_pos)
  endif
endfunction
```

**Comment by AndrewRadev** [comment]
Ah, I see, so it's kind of like, if you move beyond the boundaries of the argument, it pops it out as a `|>` chain. Nice idea. I don't write elixir myself, but I see how this could be useful. Happy to help with splitjoin bugs if you can pinpoint a specific case that doesn't work.

**Comment by sodapopcan** [comment]
Ya, it's sorta feels like just a neat party trick but using it today felt very nice.  I can't believe I didn't use sideways before because I realize I re-order parameters far more often then I think.  Usually `dt,` and `dt)` work just fine but it's become clear that it feels quite uncomfortable if the argument happens to be a list with a few commas in it.

I'm studying your work a bit and learning a lot about Vim regex.  Elixir is extremely close to Ruby and I'm trying to adapt the Ruby one to Elixir.  There are some pretty key differences in that hashes (or maps as they are calling Elixir) are delimited with `%{` and `}`, there are no block (`|`) delimiters, function definitions always end in either `do` or `, do:`... and of course there are also anonymous functions with look like `fn args ->` with no `do`.

I'll warn that I'm on a bit of a quest right now to make Elixir better for classic Vimmers since most of the community has moved on to NeoVim, although at some point I may have to give up and move on myself 😅  Not that this wouldn't still be useful in nvim, but it feels like an easier task for Treesitter?  At the same time, I've been deep into regex lately and been really surprised at what's possible and how much simpler the patterns end up being than I thought they would.

**Comment by AndrewRadev** [comment]
> Not that this wouldn't still be useful in nvim, but it feels like an easier task for Treesitter?

In theory, treesitter is perfect for a task like this. In practice, it comes with a lot of baggage. There's performance issues (including for [elixir](https://github.com/nvim-treesitter/nvim-treesitter/issues/1957)), there's stability issues (like the markdown parser [crashing the editor](https://github.com/nvim-treesitter/nvim-treesitter/issues/872)), and there's limitations (like stuff that splitjoin.vim can handle, but a treesitter-based solution [can't](https://github.com/Wansmer/treesj/issues/81)).

Some of this might just be implementation issues, but there's a fundamental problem, I feel. A "treesitter parser" is a monolithic thing -- you use what it gives you. Regexes work with whatever text you'd like to match, so they're a much more flexible tool. Their drawback is, of course, they're imprecise. Using the "skip" parameter in Vim helps a lot, but if you want to make a general-purpose "expression" text object, that'll be difficult with regexes and syntax groups and very easy with treesitter (as long as your file isn't too big :)).

I'd encourage you to try things out for yourself, if you have the patience. For me, both neovim as a whole and its treesitter integration in particular are too flaky and limited for me to switch.

**Comment by sodapopcan** [comment]
Thanks so much for your thoughts, I'm always happy to hear any amount of discussion/ranting on this topic ;)  I could also write many words on the subject thought probably much less informed than yourself.  Emotion certainly plays into it, but I don't think I will actually be switching to nvim at any point soon.  I tried a few months ago and my vimrc didn't JustWork.  After about a week of ripping out old standards and replacing with their Lua-rewrites (😢), I was in a worse-off place than I was before.  Treesitter was more interesting to me for possibly making text object stuff easier, though it's very interesting to know that there is stuff regex can do that treesitter can't!  Syntax highlighting I don't care about because I always put a little extra effort in tweaking each language.  I often get compliments on my colourschemes and last week a new coworker actually assumed I was using Treesitter!  Treesitter is certainly interesting in that it's a cross-editor solution and that one is a bit harder for me to ignore but also like, why doesn't everyone just use Vim? 🙃 🙃 🙃 

I do have the patience for regex!  I've been doing deep for the first time in years of using Vim into implementing text objects from scratch and have been learning a ton.  I've got myself a good set of very robust Elixir textobjs going that are almost ready, but still some edge cases due to the weirdness Elixir allows.  Experiencing how accurate sideways is the past few days gives me hope I can cover those last few cases that were feeling impossible.

**Comment by AndrewRadev** [comment]
>  I tried a few months ago and my vimrc didn't JustWork. After about a week of ripping out old standards and replacing with their Lua-rewrites (😢), I was in a worse-off place than I was before. 

That sounds about right, yes. In the early days of neovim, I just didn't have syntax highlighting at all, because somehow they'd borked up something terminal-related. I know a few people that experienced the same thing. These days, this works alright, but there's a lot of other issues. For instance:

- They removed `+clientserver` very early on and replaced it with something similar. That's not a bad idea, since `+clientserver` is implemented in a weird way, but the thing they replaced it with just doesn't work: https://github.com/neovim/neovim/issues/21600. This is a "known issue", but the documentation doesn't mention that at all. So why remove the original code before you have a functional replacement?
- In `:help api-indexing`, you could learn that most lua-based APIs use 0-based indexing for lines and columns... Which is at odds with any other tools that deal with lines and columns, for instance, backtraces. Vim consistently uses 1 for the first line and first column. Incidentally, *some* functions apparently use 1-based indexing for lines and 0-based for columns. I do not see an explanation in the docs about why this is the case.
- Neovim and its community are all-in on Lua 5.1, a language that has been frozen since [2012](https://www.lua.org/versions.html#5.1). Future (current) versions are backwards-incompatible. The bet, as far as I understand it, is that as long as you have a "good enough" language, the core team can just add new functions and leave the language as-is. It's difficult for me to see this working out in the long run. In 30 years, it's hard for me to imagine that Lua 5.1 would still be the pinnacle of both programming and performance, and it's hard to see how the "Neo" part of the name still applies. Vimscript might have a lot of problems, but at least it's capable of changing and adopting modern techniques, even in a clumsy way.

Basically, I have some philosophical differences with the way the project is managed, and even if I could migrate, I don't have a lot of hope for its future development. I can't predict the future, of course, but hey, my Vim setup works fine for me, so I'm not exactly in trouble. And the Vim project itself keeps slowly chugging along, as it has been for 35 years. So, whatever happens, happens.

**Comment by AndrewRadev** [comment]
Anyway, ranting aside, I'll go ahead and close this issue for now. Feel free to open a new one for follow-up problems or anything else you run into.

**Comment by sodapopcan** [comment]
Er, here I am a month later to say thanks again for all your additional thoughts even if ranty, I enjoy reading it and very informative.  And thanks again for taking care of my feature request!  I'm _still_ working on my Elixir plugin though should release soon.  I'm going to take a look at improving Elixir's piping in splitjoin so I _may_ hit up for help but I'll see what I can do on my own first.

Thanks again!  (Sorry, I'm Canadian, I have to over-thank, and of course include the word "sorry" as well 🙃).
